tests/fixtures/when.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fixtures
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	git "github.com/go-git/go-git/v5"
+	"github.com/numaproj-labs/numaplane/pkg/apis/numaplane/v1alpha1"
+	planepkg "github.com/numaproj-labs/numaplane/pkg/client/clientset/versioned/typed/numaplane/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type When struct {
+	t             *testing.T
+	restConfig    *rest.Config
+	kubeClient    kubernetes.Interface
+	gitSync       *v1alpha1.GitSync
+	gitSyncClient planepkg.GitSyncInterface
+}
+
+func (w *When) CreateGitSyncAndWait() *When {
+
+	w.t.Helper()
+	if w.gitSync == nil {
+		w.t.Fatal("No GitSync to create")
+	}
+	w.t.Log("Creating GitSync", w.gitSync.Name)
+	ctx := context.Background()
+	i, err := w.gitSyncClient.Create(ctx, w.gitSync, metav1.CreateOptions{})
+	if err != nil {
+		w.t.Fatal(err)
+	} else {
+		w.gitSync = i
+	}
+	// wait for GitSync to run
+	if err := WaitForGitSyncRunning(ctx, w.gitSyncClient, w.gitSync.Name, defaultTimeout); err != nil {
+		w.t.Fatal(err)
+	}
+	return w
+
+}
+func (w *When) DeleteGitSync() *When {
+
+	w.t.Helper()
+	if w.gitSync == nil {
+		w.t.Fatal("No gitsync to delete")
+	}
+	w.t.Log("Deleting gitsync", w.gitSync.Name)
+	ctx := context.Background()
+	if err := w.gitSyncClient.Delete(ctx, w.gitSync.Name, metav1.DeleteOptions{}); err != nil {
+		w.t.Fatal(err)
+	}
+
```

Comment:
nit: wonder if we might have symmetry between `CreateGitSyncAndWait()` and this method? If that one waits, is there any reason that this one shouldn't wait as well?

Reply:
We can achieve this similar to how Numaflow checks that a Pipeline is deleted by listing all Pods for the Pipeline and checking that the returned list size is 0.

For a GitSync, we can use the client to list them and check that this list is 0 as well.

Reply:
We just need to check that the GitSync has been deleted

Reply:
We could wait for gitSync.Get() to fail.

Reply:
Sure. I think ideally also check the error to see if it's one which indicates that the resource doesn't exist, as opposed to another failure.

Reply:
https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/controller.go#L567
`apiError.IsNotFound()`

Reply:
Sounds good, this is the current implementation now for checking GitSync and resources:

```
if err != nil {
			if errors.IsNotFound(err) {
				return w
			} else {
				w.t.Fatalf("Error getting gitSync: %v", err)
			}
		}
```